File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_Android/JavaWrapperDemo/wrapper/src/main/java/com/hyphenate/cwrapper/EMCWrapper.java
package com.hyphenate.cwrapper;

import com.hyphenate.wrapper.EMWrapper;
import com.hyphenate.wrapper.EMWrapperHelper;
import com.hyphenate.wrapper.callback.EMWrapperCallback;
import com.hyphenate.wrapper.util.EMSDKMethod;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class EMCWrapper {
    int iType = 0;
    // 用于保持callback指针，初始化时c++将自己listener的指针传过来放到java层保存，当需要是从java层取。
    long nativeListener = 0;
    static EMCWrapper cWrapper;

    EMWrapper wrapper;
    public static EMCWrapper cWrapper(int iType, long listener) {
        cWrapper = new EMCWrapper(iType, listener);
        return cWrapper;
    }

    public EMCWrapper(int iType, long listener){
        this.iType = iType;
        this.nativeListener = listener;
        wrapper = new EMWrapper();
        EMWrapperHelper.listener = new EMCWrapperListener();
    }

    public void nativeCall(String manager, String method, String jsonString, String cid) throws JSONException{
        wrapper.callSDKApi(manager, method, jsonString, new EMWrapperCallback(){
            @Override
            public void onSuccess(Object obj){
                try {
                    JSONObject jsonObject = new JSONObject();
                    if (obj != null) {
                        // value 里是 JSONObject to String
                        if (obj instanceof JSONObject || obj instanceof JSONArray) {
                            jsonObject.put("value", obj.toString());
                        }else if (obj instanceof String) {
                            JSONObject tmpJsonObject = new JSONObject();
                            tmpJsonObject.put("ret", obj);
                            jsonObject.put("value", tmpJsonObject.toString());
                        }
                    }
                    jsonObject.put("callbackId", cid);
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callback, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onError(Object obj){
                try {
                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("callbackId", cid);
                    if (obj != null) {
                        // error 里是 HyphenateException to JSONObject
                        jsonObject.put("error", obj);
                    }
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callback, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onProgress(int progress) {
                try {
                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("callbackId", cid);
                    // progress 里是 int progress
                    jsonObject.put("progress", progress);
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callbackProgress, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        });
    }

    public String nativeGet(String manager, String method, String jsonString, String cid) {
        System.out.println("java: nativeGet: manager:" + manager + " method:" + method + " jsonString:" + jsonString + " cid:" + cid);
        String str = wrapper.callSDKApi(manager, method, jsonString, new EMWrapperCallback(){
            @Override
            public void onSuccess(Object obj){
                try {
                    JSONObject jsonObject = new JSONObject();
                    if (obj != null) {
                        jsonObject.put("value", obj);
                    }
                    jsonObject.put("callbackId", cid);
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callback, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onError(Object obj){
                try {
                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("callbackId", cid);
                    if (obj != null) {
                        jsonObject.put("error", obj);
                    }
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callback, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onProgress(int progress) {
                try {
                    JSONObject jsonObject = new JSONObject();
                    jsonObject.put("callbackId", cid);
                    jsonObject.put("progress", progress);
                    EMWrapperHelper.listener.onReceive(EMSDKMethod.callbackProgress, cid, jsonObject.toString());
                }catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        });

        return str;
    }
}

File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_iOS/ChatCWrapper/ChatCWrapper/src/EMCWrapper.m
//
//  EMCWrapper.m
//  ChatCWrapper
//
//  Created by 杜洁鹏 on 2022/11/16.
//

#import "EMCWrapper.h"
#import <wrapper/wrapper.h>
#import "EMCWrapperListener.h"
#import "NSString+Category.h"

@interface EMCWrapper ()
@property (nonatomic, strong) EMWrapper *wrapper;
@end

@implementation EMCWrapper
{
    void* callback;
    int iType;
}

- (instancetype)initWithType:(int)iType
                    listener:(void*)listener {
    if (self = [super init]) {
        self.wrapper = [[EMWrapper alloc] init];
        EMWrapperHelper.shared.listener = [[EMCWrapperListener alloc] init:listener];
    }
    return self;
}

- (NSString *)nativeGet:(const char *)manager
                 method:(const char *)method
                 params:(const char*)jstr
                    cid:(const char *)cbid {
    EMWrapperCallback *callback = [[EMWrapperCallback alloc] init];
    __block NSString *callbackId = [NSString fromChar:cbid];
    callback.onSuccessCallback = ^(NSObject *valueObj) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        if (valueObj != nil) {
            dict[@"value"] = valueObj;
        }
        dict[@"callbackId"] = callbackId;
        [EMWrapperHelper.shared.listener onReceive:@"callback" method:callbackId info:[dict toJsonString]];
    };
    callback.onErrorCallback = ^(NSDictionary *errorDict) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        dict[@"callbackId"] = [NSString fromChar:cbid];
        dict[@"error"] = errorDict;
        [EMWrapperHelper.shared.listener onReceive:@"callback" method:callbackId info:[dict toJsonString]];
    };
    callback.onProgressCallback = ^(int progress) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        dict[@"callbackId"] = [NSString fromChar:cbid];
        dict[@"progress"] = @(progress);
        [EMWrapperHelper.shared.listener onReceive:@"callbackProgress" method:callbackId info:[dict toJsonString]];
    };
    
    NSString *ret = [self.wrapper callSDKApi:[NSString fromChar:manager]
                                      method:[NSString fromChar:method]
                                      params:[NSString fromChar:jstr].toDict
                                    callback:callback];
    return ret;
}

- (void)nativeCall:(const char *)manager
            method:(const char *)method
            params:(const char*)jstr
               cid:(const char *)cbid
{
    __block NSString *callbackId = [NSString fromChar:cbid];
    EMWrapperCallback *callback = [[EMWrapperCallback alloc] init];
    callback.onSuccessCallback = ^(NSObject *valueObj) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        if (valueObj != nil) {
            if ([valueObj isKindOfClass:[NSDictionary class]]) {
                dict[@"value"] = [(NSDictionary *)valueObj toJsonString];
            }else if ([valueObj isKindOfClass:[NSArray class]]) {
                dict[@"value"] = [(NSArray *)valueObj toJsonString];
            }
        }
        dict[@"callbackId"] = callbackId;
        [EMWrapperHelper.shared.listener onReceive:@"callback" method:callbackId info:[dict toJsonString]];
    };
    callback.onErrorCallback = ^(NSDictionary *errorDict) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        dict[@"callbackId"] = [NSString fromChar:cbid];
        dict[@"error"] = errorDict;
        [EMWrapperHelper.shared.listener onReceive:@"callback" method:callbackId info:[dict toJsonString]];
    };
    callback.onProgressCallback = ^(int progress) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        dict[@"callbackId"] = [NSString fromChar:cbid];
        dict[@"progress"] = @(progress);
        [EMWrapperHelper.shared.listener onReceive:@"callbackProgress" method:callbackId info:[dict toJsonString]];
    };
    
    [self.wrapper callSDKApi:[NSString fromChar:manager]
                                  method:[NSString fromChar:method]
                                  params:[NSString fromChar:jstr].toDict
                                callback:callback];
}
@end
